### PR TITLE
refactor: replace String with Tenant

### PR DESCRIPTION
**Pull request by drmingdrmer**
I hereby agree to the terms of the CLA available at: https://docs.databend.com/dev/policies/cla/

##### refactor: replace String with Tenant

Replacing all String with Tenant introduces a lot changes.
These changes will be split into several PRs.
This is one of them.

- Part of #14719

## Tests

- [x] Unit Test
- [ ] Logic Test
- [ ] Benchmark Test
- [ ] No Test  - _Explain why_

## Type of change

- [x] Refactoring

## Related Issues

- - -
This change is [Reviewable](https://reviewable.io/reviews/datafuselabs/databend/15119)

### CODE REVIEW

File: src/query/ast/src/parser/statement.rs [hunk 1]
```diff
@@ -3611,18 +3612,30 @@ pub fn create_database_option(i: Input) -> IResult<CreateDatabaseOption> {
         rule! {
             FROM ~ SHARE ~ #ident ~ "." ~ #ident
         },
-        |(_, _, tenant, _, share_name)| {
+        |(_x, _y, tenant, _z, share_name)| {
```
Comment (andylokandy) [inline src/query/ast/src/parser/statement.rs:3615]: why xyz?

File: src/query/ast/src/parser/statement.rs [hunk 2]
```diff
@@ -3611,18 +3612,30 @@ pub fn create_database_option(i: Input) -> IResult<CreateDatabaseOption> {
...
             CreateDatabaseOption::FromShare(ShareNameIdent {
-                tenant: tenant.to_string(),
+                // TODO: non-safe new() that does not return an error
+                tenant: Tenant::new(tenant.to_string()),
```
Comment (andylokandy) [inline src/query/ast/src/parser/statement.rs:3618]: You can use `map_res` to return an error.

File: src/query/ast/src/parser/statement.rs [hunk 3]
```diff
@@ -3611,18 +3612,30 @@ pub fn create_database_option(i: Input) -> IResult<CreateDatabaseOption> {
...
+    if let CreateDatabaseOption::FromShare(ident) = &opt {
+        if ident.tenant.name().is_empty() {
+            return Err(nom::Err::Error(Error::from_error_kind(
+                i,
+                ErrorKind::Other("tenant is empty string"),
+            )));
+        }
+    }
+
+    Ok((i, opt))
```
Comment (andylokandy) [inline src/query/ast/src/parser/statement.rs:3638]: Use `map_res` instead.

**Comment by drmingdrmer** [comment]
@andylokandy 
Thank you for your advices. I'm gonna fix these issues in next PR

**Pull request merged**